notebooks/Threshold Optimization Postprocessing for Binary Classification.ipynb

```diff
@@ -0,0 +1,458 @@
+    "\n",
+    "The more interesting logic happens in between. The x and y values were calculated as follows:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "n_group_0 = {}\n",
+    "n_group_1 = {}\n",
+    "for group_name, group in data_grouped_by_attribute:\n",
+    "    print(\"{}:\".format(group_name))\n",
+    "    n_group_1[group_name] = sum(group[\"label\"])\n",
+    "    n_group_0[group_name] = len(group) - n_group_1[group_name]\n",
+    "    \n",
+    "    print(\"    number of samples with label 1: {}\".format(n_group_1[group_name]))\n",
+    "    print(\"    number of samples with label 0: {}\".format(n_group_0[group_name]))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "threshold = 0.5\n",
+    "for group_name, group in data_grouped_by_attribute:\n",
+    "    x_group_0_5 = sum((group[\"score\"] > threshold) & (group[\"label\"] == 0)) / n_group_0[group_name]\n",
+    "    y_group_0_5 = sum((group[\"score\"] > threshold) & (group[\"label\"] == 1)) / n_group_1[group_name]\n",
+    "    print(\"{}:\".format(group_name))\n",
+    "    print(\"    P[Ŷ = 1 | Y = 0] = {}\".format(x_group_0_5))\n",
+    "    print(\"    P[Ŷ = 1 | Y = 1] = {}\".format(y_group_0_5))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that it never makes sense to have $x>y$ because in that case you're better off flipping labels, i.e. completely turning around the meaning of the scores. Our method automatically does that unless specified otherwise.\n",
+    "\n",
+    "## Interpolated Predictions and Probabilistic Classifiers\n",
+    "\n",
+    "This way you end up with a set of points above the diagonal line connecting (0,0) and (1,1). We calculate the convex hull based on that, because we can reach any point in between two known thresholding points by interpolation. An interpolation could be $p_0 (x_0, y_0) + p_1 (x_1, y_1)$. For our post processing algorithm that would mean that we use the rule defined by $(x_0, y_0, \\text{operation}_0)$ $\\quad p_0$ percent of the time, and the rule defined by $(x_1, y_1, \\text{operation}_1)$ $\\quad p_1$ percent of the time, thus resulting in a probabilistic classifier. Depending on the data certain fairness objectives can only be accomplished with probabilistic classifiers. However, not every use case lends itself to probabilistic classifiers, since it could mean that two people with identical features are classified differently.\n",
+    "\n",
+    "## Finding the Equalized Odds solution\n",
+    "\n",
+    "In this specific instance we don't need to do any extra work to get the convex hull, since there's only one point above the diagonal line."
```

Comment:
"only one point above the diagonal line" ? It looks like both curves are entirely above the diagonal line.

Reply:
Of course, the "diagonal line" is not plotted, and the lines are not labelled either.

Reply:
And there are plenty of lines one could draw above the diagonal which would still require work to get the convex hull. Surely not having any points below the diagonal just means that you would be flipping the point?

Reply:
We don't do points below the line for that reason. I changed the description.